torchaudio/prototype/models/hdemucs.py

```diff
@@ -0,0 +1,939 @@
+"""
+This code contains the spectrogram and Hybrid version of Demucs.
+"""
+import math
+import typing as tp
+from typing import List, Optional
+
+import torch
+from torch import nn
+from torch.nn import functional as F
+
+
+class ScaledEmbedding(torch.nn.Module):
+    r"""Make continuous embeddings and boost learning rate
+
+    Args:
+        num_embeddings (int): number of embeddings
+        embedding_dim (int): embedding dimensions
+        scale (float): amount to scale (Default: 10.0)
+        smooth (bool): choose to apply smoothing (Default: ``False``)
+    """
+
+    def __init__(self, num_embeddings: int, embedding_dim: int, scale: float = 10.0, smooth: bool = False):
+        super().__init__()
+        self.embedding = nn.Embedding(num_embeddings, embedding_dim)
+        if smooth:
+            weight = torch.cumsum(self.embedding.weight.data, dim=0)
+            # when summing gaussian, overscale raises as sqrt(n), so we nornalize by that.
+            weight = weight / torch.arange(1, num_embeddings + 1).to(weight).sqrt()[:, None]
+            self.embedding.weight.data[:] = weight
+        self.embedding.weight.data /= scale
+        self.scale = scale
+
+    @property
+    def weight(self) -> torch.Tensor:
+        return self.embedding.weight * self.scale
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        r"""Forward pass for embedding with scale.
+        Args:
+            x (torch.Tensor): input tensor .
+
+        Returns:
+            (torch.Tensor):
+                Embedding output
+        """
+
+        out = self.embedding(x) * self.scale
+        return out
+
+
+class HEncLayer(torch.nn.Module):
+    def __init__(
+        self,
+        chin: int,
+        chout: int,
+        kernel_size: int = 8,
+        stride: int = 4,
+        norm_groups: int = 1,
+        empty: bool = False,
+        freq: bool = True,
+        norm: bool = True,
+        context: int = 0,
+        dconv_kw: dict = {},
+        pad: bool = True,
+    ):
+        r"""Encoder layer. This used both by the time and the frequency branch.
+        Args:
+            chin (int): number of input channels.
+            chout (int): number of output channels.
+            kernel_size (int): Kernel size for encoder (Default: 8)
+            stride (int): Stride for encoder layer (Default: 4)
+            norm_groups (int): number of groups for group norm. (Default: 11)
+            empty: used to make a layer with just the first conv. this is used
+                before merging the time and freq. branches. (Default: ``False``)
+            freq (bool): boolean for whether conv layer is for frequency (Default: ``True``)
+            norm (bool): use GroupNorm. (Default: ``True``)
```

Comment:
same here, we can change it to "norm_type", the options can be `["none", "group_norm"]`.